### ISSUE TITLE
async hooks are not restarted when a property change.

### ISSUE THREAD

**Issue by rambip** [DioxusLabs/dioxus#2784]
**Problem**

When a component has a non-signal input and use `use_future` or `use_resource`, the behavior is odd: the component is not re-rendered.

**Steps To Reproduce**

- create a new dioxus project with the web render.
- add `async_std` dependency
- use this code in main.rs:

```rust
#![allow(non_snake_case)]

use dioxus::prelude::*;
use tracing::{Level, info};

use async_std::task;
use std::time::Duration;


fn main() {
    dioxus_logger::init(Level::INFO).expect("failed to init logger");
    launch(App);
}

async fn request(word: String) -> String {
    info!("request made");
    task::sleep(Duration::from_secs(1)).await;
    format!("I have some {word}s")
}


///panicked at /home/rambip/.cargo/registry/src/index.crates.io-6f17d22bba15001f/dioxus-core-0.5.1/src/diff/node.rs:113:18:
// not yet implemented: This is an usual custom case for dynamic nodes. We don't know how to handle it yet.
#[component]
fn Greeter1(word: String) -> Element {
    info!("greeter 1 called");
    let view = use_resource(move || request(word.clone()));

    rsx!{
        { view }
    }
}


// not updated after the first change
#[component]
fn Greeter2(word: String) -> Element {
    info!("greeter 2 called");
    let view = use_resource(move || request(word.clone()));

    rsx!{
        match &*view.read() {
            Some(x) => {x},
            None => "wait ...",
        }
    }
}

#[component]
fn App() -> Element {
    let mut fruit: Signal<Option<String>> = use_signal(|| None);

    rsx! {
        button {
            onclick: move |_| *fruit.write() = Some("apple".to_string()),
            "apple",
        },
        button {
            onclick: move |_| *fruit.write() = Some("banana".to_string()),
            "banana",
        },
        hr {}
        if let Some(x) = &*fruit.read() {
            Greeter1 {
                word: x,
            }
        }
    }
}

```
- run `dx serve`
- open the browser, open the console and click on a button. Change `Greeter1` to `Greeter2` to see the difference.

You should see an error message for `Greeter1` and a component not re-rendering for `Greeter2`.

It gets even worse for this code:

```rust
#![allow(non_snake_case)]

use dioxus::prelude::*;
use tracing::{Level, info};

use async_std::task;
use std::time::Duration;


fn main() {
    dioxus_logger::init(Level::INFO).expect("failed to init logger");
    launch(App);
}

async fn request(word: String) -> String {
    info!("request made");
    task::sleep(Duration::from_secs(1)).await;
    format!("I have some {word}s")
}


// not updated after the first change
fn greeter(word: String) -> Element {
    let view = use_resource(move || request(word.clone()));

    rsx!{
        match &*view.read() {
            Some(x) => {x},
            None => "wait ...",
        }
    }
}

#[component]
fn App() -> Element {

    let mut view = use_signal(|| None);

    rsx! {
        button {
            onclick: move |_| *view.write() = greeter("apple".to_string()),
            "apple",
        },
        button {
            onclick: move |_| *view.write() = greeter("banana".to_string()),
            "banana",
        },
        hr {}
        {view}
    }
}
```

This time, the component stays stuck at "wait ...".

You get a slightly different behavior if you declare `Greeter` as a component instead of a function.

**Expected behavior**

In each one of these cases, I would expect the future to restart and the component to re-render every time the input changes, even if it is not a signal.

**Environment:**
 - Dioxus version:  0.5.1
 - App platform:    web

**Questionnaire**

- [x] I'm interested in fixing this myself but don't know where to start
- [ ] I would like to fix and I have a solution
- [ ] I don't have time to fix this right now, but maybe later

**Comment by marc2332** [comment]
I think this already got fixed a while ago, can you try with Dioxus 0.5.6?

**Comment by rambip** [comment]
I get exactly the same errors

**Comment by ealmloff** [comment]
Dioxus cannot track non reactive values. You can either make your value reactive by accepting `ReadOnlySignal<T>` instead of `T` or use the `use_reactive` hook to create a closure that is reactive with explicit dependencies. Both options are documented in the new 0.6.0-alpha docs [here](https://docs.rs/dioxus-hooks/0.6.0-alpha.0/dioxus_hooks/fn.use_resource.html#with-non-reactive-dependencies).

The diffing bug was fixed in #2746 

Here is a fixed version of the code you posted:
```rust
#![allow(non_snake_case)]

use dioxus::prelude::*;

use async_std::task;
use std::time::Duration;

fn main() {
    launch(App);
}

async fn request(word: String) -> String {
    task::sleep(Duration::from_secs(1)).await;
    format!("I have some {word}s")
}

// You can make word reactive by wrapping it in ReadOnlySignal
#[component]
fn Greeder(word: ReadOnlySignal<String>) -> Element {
    let view = use_resource(move || request(word()));

    rsx! {
        match &*view.read() {
            Some(x) => {x},
            None => "wait ...",
        }
    }
}

#[component]
fn App() -> Element {
    let mut view = use_signal(|| rsx! {});

    rsx! {
        button {
            // You can't call greeder directly because it would violate the rules of hooks
            // (https://docs.rs/dioxus-hooks/0.6.0-alpha.0/dioxus_hooks/fn.use_resource.html#additional-information-that-may-be-useful) 
            // But you can set rsx to the Greeder component
            onclick: move |_| *view.write() = rsx! { Greeder { word: "apple" } },
            "apple"
        }
        button {
            onclick: move |_| *view.write() = rsx! { Greeder { word: "banana" } },
            "banana"
        }
        hr {}
        {view}
    }
}
```

Eventually we should add a warning when you use non-reactive props directly in reactive hooks in dx check (with #1460)
